src/ray/core_worker/task_event_buffer.cc

```diff
@@ -351,33 +418,145 @@ void TaskEventBufferImpl::FlushEvents(bool forced) {
-                       << status.ToString() << "]";
+      on_failed_to_send_to_gcs(status);
+    } else {
+      stats_counter_.Increment(kTotalNumTaskAttemptsReported, num_task_attempts_to_send);
+      stats_counter_.Increment(kTotalNumLostTaskAttemptsReported,
+                               num_dropped_task_attempts_to_send);
+      stats_counter_.Increment(kTotalTaskEventsBytesReported, num_bytes_to_send);
     }
     grpc_in_progress_ = false;
   };
 
   auto status = task_accessor->AsyncAddTaskEventData(std::move(data), on_complete);
-  {
-    if (!status.ok()) {
-      // If we couldn't even send the data by invoking client side callbacks, there's
-      // something seriously wrong, and losing data in this case should not be too
-      // worse. So we will silently drop these task events.
-      RAY_LOG(WARNING)
-          << "Failed to push task state events to GCS. Data will be lost. [status="
-          << status.ToString() << "]";
-      grpc_in_progress_ = false;
-
-      // Fail to send, currently dropping events.
-      stats_counter_.Increment(TaskEventBufferCounter::kTotalNumTaskProfileEventDropped,
-                               num_profile_event_to_send);
-      stats_counter_.Increment(TaskEventBufferCounter::kTotalNumTaskStatusEventDropped,
-                               num_status_event_to_send);
+  if (!status.ok()) {
+    on_failed_to_send_to_gcs(status);
+    grpc_in_progress_ = false;
+  }
+}
+
+void TaskEventBufferImpl::ResetCountersForFlush() {
+  // Profile events dropped.
+  auto num_profile_events_dropped_since_last_flush = stats_counter_.Get(
+      TaskEventBufferCounter::kNumTaskProfileEventDroppedSinceLastFlush);
+  stats_counter_.Decrement(
+      TaskEventBufferCounter::kNumTaskProfileEventDroppedSinceLastFlush,
+      num_profile_events_dropped_since_last_flush);
+  stats_counter_.Increment(TaskEventBufferCounter::kTotalNumTaskProfileEventDropped,
+                           num_profile_events_dropped_since_last_flush);
+
+  // Task events dropped.
+  auto num_status_events_dropped_since_last_flush = stats_counter_.Get(
+      TaskEventBufferCounter::kNumTaskStatusEventDroppedSinceLastFlush);
+  stats_counter_.Decrement(
+      TaskEventBufferCounter::kNumTaskStatusEventDroppedSinceLastFlush,
+      num_status_events_dropped_since_last_flush);
+  stats_counter_.Increment(TaskEventBufferCounter::kTotalNumTaskStatusEventDropped,
+                           num_status_events_dropped_since_last_flush);
+}
+
+void TaskEventBufferImpl::AddTaskEvent(std::unique_ptr<TaskEvent> task_event) {
+  if (task_event->IsProfileEvent()) {
+    AddTaskProfileEvent(std::move(task_event));
+  } else {
+    AddTaskStatusEvent(std::move(task_event));
+  }
+}
+
+void TaskEventBufferImpl::AddTaskStatusEvent(std::unique_ptr<TaskEvent> status_event) {
+  absl::MutexLock lock(&mutex_);
+  if (!enabled_) {
+    return;
+  }
+
+  if (dropped_task_attempts_since_last_flush_.count(status_event->GetTaskAttempt())) {
```

Comment:
Same question as below, but is this necessary for performance? Can we just not do this and handle this in GCS?

Reply:
We could - but this would result in more "wasted" data being sent. 

A single task attempt could generate:
- 8 status change updates + 8 profile events 

If one of the status change event gets dropped on the worker, the entire task should be dropped and GCed since we are reporting data loss at task attempt level now. 

Sending/recording the remaining 15 updates should probably be avoided if it doesn't make things too complicated?